x-pack/plugins/gis/public/components/widget_overlay/layer_control/layer_toc/toc_entry/view.js

```diff
@@ -32,63 +28,30 @@ export class TOCEntry extends React.Component {
     this._isMounted = false;
   }
 
-  _renderVisibilityToggle() {
-    const { layer, toggleVisible } = this.props;
-    return (
-      <VisibilityToggle
-        id={layer.getId()}
-        checked={layer.isVisible()}
-        onChange={() => toggleVisible(layer.getId())}
-        size={'l'}
-      >
-        {layer.getIcon()}
-      </VisibilityToggle>
-    );
+
+  async _updateDisplayName() {
+    const label = await this.props.layer.getDisplayName();
+    if (this._isMounted) {
+      if (label !== this.state.displayName) {
+        this.setState({
+          displayName: label
+        });
+      }
+    }
   }
 
-  render() {
+  componentDidUpdate() {
+    this._updateDisplayName();
+  }
 
-    const { layer, openLayerPanel, zoom } = this.props;
+  render() {
 
-    const displayName = layer.getDisplayName();
-    Promise.resolve(displayName).then(label => {
-      if (this._isMounted) {
-        if (label !== this.state.displayName) {
-          this.setState({
-            displayName: label
-          });
-        }
-      }
+    const { layer, openLayerPanel, zoom, toggleVisible, fitToBounds } = this.props;
+    const legendIcon = layer.renderTocEntryIcon({
```

Comment:
Same comment as above. Why bury the react components behind the function call `layer.renderTocEntryIcon`. Just do that inline here so there is less layers of indirection

```
const legendIcon = (
      <LayerTocActions
        layer={layer}
        fitToBounds={() => {
          fitToBounds(layer.getId());
        }}
        zoom={zoom}
        toggleVisible={() => {
          toggleVisible(layer.getId());
        }}
      />
    );
```

Reply:
good call!